vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd

```diff
@@ -0,0 +1,969 @@
+```
+
+In addition, the `loss` property also contains regularization losses created
+for the weights of any inner layer:
+
+```{r}
+OuterLayerWithKernelRegularizer(keras$layers$Layer) %py_class% {
+  initialize <- function() {
+    super$initialize()
+    self$dense <- layer_dense(units = 32, kernel_regularizer = regularizer_l2(1e-3))
+  }
+  call <- function(inputs) {
+    self$dense(inputs)
+  }
+}
+
+layer <- OuterLayerWithKernelRegularizer()
+layer(tf$zeros(shape(1, 1))) |> invisible()
+
+# This is `1e-3 * sum(layer$dense$kernel ** 2)`,
+# created by the `kernel_regularizer` above.
+print(layer$losses)
+```
+
+These losses are meant to be taken into account when writing training loops,
+like this:
+
+```{r}
+zip <- function(..., simplify = TRUE) {
+  .mapply(if (simplify) c else list, list(...), NULL)
+}
+```
+
+```{r, eval=FALSE}
+# Instantiate an optimizer.
+optimizer <- optimizer_sgd(learning_rate = 1e-3)
+loss_fn <- loss_sparse_categorical_crossentropy(from_logits = TRUE)
+
+# while(!is.null(batch <- iter_next(dataset_iterator))) { ... } ## too compact?
+# batch <- iter_next(dataset_iterator) %\\% break               ## depend rlang?
+# autograph({ for (batch in dataset) ...})                      ## too magical?
+# ag_for(batch, dataset, {...})                                 ## too unusual?
+
+# Iterate over the batches of a dataset.
+# train_dataset <- dataset_mnist()
+dataset_iterator <- reticulate::as_iterator(train_dataset)
+repeat {
+  batch <- reticulate::iter_next(dataset_iterator)
+  if (is.null(batch)) break
+  c(x_batch_train, y_batch_train) %<-% batch
+  with(tf$GradientTape() %as% tape, {
+    logits <- layer(x_batch_train) # Logits for this minibatch
+    # Loss value for this minibatch
+    loss_value <- loss_fn(y_batch_train, logits)
+    # Add extra losses created during this forward pass:
+    loss_value <- loss_value + sum(model$losses)
+  })
+  grads <- tape$gradient(loss_value, model$trainable_weights)
+  optimizer$apply_gradients(zip(grads, model$trainable_weights))
+}
+```
+
+For a detailed guide about writing training loops, see the
+[guide to writing a training loop from scratch](/guides/writing_a_training_loop_from_scratch/).
+
+These losses also work seamlessly with `fit()` (they get automatically summed
+and added to the main loss, if any):
+
+```{r, eval=TRUE}
+input <- layer_input(shape(3))
+output <- input %>% layer_activity_regularization()
+# output <- ActivityRegularizationLayer()(input)
+model <- keras_model(input, output)
+
+# If there is a loss passed in `compile`, the regularization
+# losses get added to it
+model %>% compile(optimizer = "adam", loss = "mse")
+model %>% fit(k_random_uniform(c(2, 3)),
+  k_random_uniform(c(2, 3)),
+  epochs = 1, verbose = FALSE
+)
+
+# It's also possible not to pass any loss in `compile`,
+# since the model already has a loss to minimize, via the `add_loss`
+# call during the forward pass!
+model %>% compile(optimizer = "adam")
```

Comment:
so great this works now! Remember i needed a workaround in tfprobability, in the past...